src/dinit-iostream.cc

```diff
@@ -0,0 +1,744 @@
+            // buffer's data in streambuf is splited into two parts in this case because of cpbuffer nature.
+            bufv[1].iov_base = buf->get_buf_base();
+            bufv[1].iov_len = buf->get_length() - bufv[0].iov_len;
+        }
+        else {
+            bufv[1].iov_base = nullptr;
+            bufv[1].iov_len = 0;
+        }
+        bufv[2].iov_base = const_cast<char *>(msg);
+        bufv[2].iov_len = remaining_bytes;
+
+        unsigned buffer_length = bufv[0].iov_len + bufv[1].iov_len;
+
+        ssize_t cur_written_bytes = bp_sys::writev(fd, bufv, 3);
+        if (cur_written_bytes < 0) {
+            io_error = errno;
+            // Some bytes may be written in previous iteration
+            return (count - remaining_bytes > 0) ? count - remaining_bytes : -1;
+        }
+        if (cur_written_bytes >= buffer_length) {
+            buf->consume(buffer_length);
+            cur_written_bytes -= buffer_length;
+            remaining_bytes -= cur_written_bytes;
+            msg += cur_written_bytes;
+        }
+        else {
+            buf->consume(cur_written_bytes);
+        }
+
+        if (remaining_bytes == 0) {
+            return count;
+        }
+    }
+
+    buf->append(msg, remaining_bytes);
+    return count;
+}
+
+void ostream::throw_exception_on(const int states)
+{
+    if ((states & io_states::buffer_failbit) && buffer_failure()) {
+        throw iostream_internal_err("Cannot use stream's buffer: buffer is nullptr!");
+    }
+    if ((states & io_states::io_failbit) && io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
+}
+
+bool ostream::open_nx(const char *path) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool ostream::open(const char *path)
+{
+    bool r = open_nx(path);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::open_nx(const char *path, const int flags) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY | flags);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool ostream::open(const char *path, const int flags)
+{
+    bool r = open_nx(path, flags);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::open_nx(const char *path, const int flags, const mode_t mode) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY | flags, mode);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool ostream::open(const char *path, const int flags, const mode_t mode)
+{
+    bool r = open_nx(path, flags, mode);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::close_nx() noexcept
+{
+    if (!flush_nx()) {
+        // io_error was set by flush_nx() call
+        return false;
```

Comment:
The file descriptor must still be closed even if flush fails.

Reply:
And what about `flush_nx()` failure? What should this function return when `flush_nx()` fails but close succeeds? `false`?

Reply:
Yes - if flush_nx() fails, there's no need to check the return of `close()`, but you must still call it. Otherwise there's no way (at least no easy way) for the caller to know if the descriptor was closed or not.

Reply:
You said there is no need for checking the result of `close()` in case of `flush_nx()` failure so you suggest that we don't need to set `io_error` on `close()` failure? It needs checking `close()`'s return.

Reply:
> so you suggest that we don't need to set io_error on close() failure?

if flush_nx() already failed then io_error is already set